src/Field.sol

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+library Field {
+    uint256 constant ZERO = 0;
+    uint256 constant ONE = 1;
+
+    /// @dev Compute f^-1 for f \in Fr scalar field
+    /// @notice credit: Aztec, Spilsbury Holdings Ltd
+    function invert(uint256 _x, uint256 _mod) public view returns (uint256 output) {
+        assembly {
+            let mPtr := mload(0x40)
+            mstore(mPtr, 0x20)
+            mstore(add(mPtr, 0x20), 0x20)
+            mstore(add(mPtr, 0x40), 0x20)
+            mstore(add(mPtr, 0x60), _x)
+            mstore(add(mPtr, 0x80), sub(_mod, 2))
+            mstore(add(mPtr, 0xa0), _mod)
+            if iszero(staticcall(gas(), 0x05, mPtr, 0xc0, 0x00, 0x20)) {
+                revert(0, 0)
+            }
+            output := mload(0x00)
+        }
+    }
+
+    function fieldpow(uint256 _base, uint256 _exp, uint256 _mod) public view returns (uint256 result) {
+        assembly {
+            // Free memory pointer
+            let pointer := mload(0x40)
+            // Define length of base, exponent and modulus. 0x20 == 32 bytes
+            mstore(pointer, 0x20)
+            mstore(add(pointer, 0x20), 0x20)
+            mstore(add(pointer, 0x40), 0x20)
+            // Define variables base, exponent and modulus
+            mstore(add(pointer, 0x60), _base)
+            mstore(add(pointer, 0x80), _exp)
+            mstore(add(pointer, 0xa0), _mod)
+            // Store the result
+            let value := mload(0xc0)
+            // Call the precompiled contract 0x05 = bigModExp
+            if iszero(staticcall(not(0), 0x05, pointer, 0xc0, value, 0x20)) {
+                revert(0, 0)
+            }
+            result := mload(value)
+        }
+    }
+
+    // @dev Perform a modular exponentiation.
+    // @return base^exponent (mod modulus)
+    // This method is ideal for small exponents (~64 bits or less), as it is cheaper than using the pow precompile
+    // @notice credit: credit: Aztec, Spilsbury Holdings Ltd
+    function powSmall(
+        uint256 base,
+        uint256 exponent,
+        uint256 modulus
+    ) internal pure returns (uint256) {
+        uint256 result = 1;
+        uint256 input = base;
+        uint256 count = 1;
+
+        assembly {
+            let endPoint := add(exponent, 0x01)
+            for {
+
+            } lt(count, endPoint) {
+                count := add(count, count)
+            } {
+                if and(exponent, count) {
+                    result := mulmod(result, input, modulus)
+                }
+                input := mulmod(input, input, modulus)
+            }
+        }
+
+        return result;
+    }
+
+    function sqrt(uint256 _x, uint256 _mod) public view returns (uint256) {
```

Comment:
It would be great to have a few comments in there including:
- naming the Tonelli-Shanks,
- naming the Legendre symbol,
- commenting on the conditions of `_mod` (primality) that needs to hold